### PR TITLE
Retain spaces in quoted strings in window options

### PR DESCRIPTION
**Pull request by kaushalmodi**
Fixes https://github.com/tony/tmuxp/issues/239

With the below in tmux config:

    setw -g pane-border-format " #P "

Before the fix:

    ('pane-border-format', '"', '#P', '"')

After the fix:

    ('pane-border-format', ' #P ')

### CODE REVIEW

**Comment by tony** [comment]
I added a test you can work again, but shlex.split isn't storing the expected value.

```
_____________________________ test_set_show_window_options ______________________________

session = Session($1 libtmux_ziQLip)

    def test_set_show_window_options(session):
        """Set option then Window.show_window_options(key)."""
        window = session.new_window(window_name='test_window')

        window.set_window_option('main-pane-height', 20)
        assert window.show_window_options('main-pane-height') == 20

        window.set_window_option('pane-border-format', ' #P ')
>       assert window.show_window_options('pane-border-format') == [' #P ']
E       assert '"' == [' #P ']
E        +  where '"' = <bound method Window.show_window_options of Window(@2 2:test_window, Session($1 libtmux_ziQLip))>('pane-border-format')
E        +    where <bound method Window.show_window_options of Window(@2 2:test_window, Session($1 libtmux_ziQLip))> = Window(@2 2:test_window, Session($1 libtmux_ziQLip)).show_w
indow_options
```

**Comment by tony** [comment]
I also rebased it against the latest changes, and dropped python 2.6

**Comment by tony** [comment]
Can you give it a look @kaushalmodi?

**Comment by kaushalmodi** [comment]
Will have a look.. could probably mean that `set_window_option` has somehow to be updated to support spaces in quotes.

**Comment by kaushalmodi** [comment]
I am missing out something obvious here.. how do I run the `test_set_show_window_options` test locally?

I tried `make test` in the libtmux repo but the `py.test` referenced in the `Makefile` does not exist.

```
> make test
py.test
make: py.test: Command not found
make: *** [test] Error 127
```

**Comment by tony** [comment]
Did you try loading the project with `tmuxp load path/to/libtmux`?  You can also manually invoke `./bootstrap_env.py` and source into the virtual environment created.

Also you can do `pip install requirements/test.txt`

**Comment by kaushalmodi** [comment]
I got this from the `.travis.yml` and it ran the tests: `python3 setup.py test`... though fails a different test:

```
_____________________________________________________________ test_set_height _____________________________________________________________

session = Session($1 libtmux_8kct1jov)

    def test_set_height(session):
        window = session.new_window(window_name='test_set_height')
        window.split_window()
        pane1 = window.attached_pane
        pane1_height = pane1['pane_height']

        pane1.set_height(4)
        assert pane1['pane_height'] != pane1_height
>       assert int(pane1['pane_height']) == 4
E       AssertionError: assert 3 == 4
E        +  where 3 = int('3')

tests/test_pane.py:36: AssertionError
================================================== 1 failed, 77 passed in 16.67 seconds ===================================================
```

**Comment by tony** [comment]
Rerun it and see if it keeps happening. Could be a fluke.

**Comment by kaushalmodi** [comment]
Still getting the same error.. will tackle this tomorrow. Thanks for the guidance.

**Comment by tony** [comment]
Yes go ahead, and try to do it via `tmuxp load path/to/libtmux` and see what happens

**Comment by kaushalmodi** [comment]
I still see that test_set_height error. But I also see the test_set_show_window_options error too.

image

**Comment by kaushalmodi** [comment]
I'll need you help to understand this problem.. here is my debug attempt: https://github.com/kaushalmodi/libtmux/commit/5e94ad8a066314e3be24092135a8a067bce2abae

Running

```
make test='tests/test_window.py::test_set_show_window_options'
```

gives

```
        window.set_window_option('pane-border-format', 'foo')
        assert window.show_window_options('pane-border-format') == '"foo"'

        window.set_window_option('pane-border-format', '#P')
        assert window.show_window_options('pane-border-format') == '"#P"'

        # window.set_window_option('pane-border-format', '#P')
        # assert window.show_window_options('pane-border-format') == ['#P']

        window.set_window_option('pane-border-format', ' #P ')
>       assert window.show_window_options('pane-border-format') == '" #P "'
E       assert '"' == '" #P "'
E         - "
E         + " #P "

tests/test_window.py:187: AssertionError
---------------------------------------------------------- Captured stdout call -----------------------------------------------------------
=== Setting Window Option ===
set_w_opt_dbg: "20"
cmd_dbg1: set-window-option (u'-t$1:2', u'main-pane-height', 20) {}
cmd_dbg2: set-window-option (u'-t$1:2', u'main-pane-height', 20) {}
=== Showing Window Option ===
show_w_opt_dbg: here1
show_w_opt_dbg: here2
show_w_opt_dbg: here4 main-pane-height False
cmd_dbg1: show-window-options (u'main-pane-height',) {}
cmd_dbg2: show-window-options (u'-t', u'@2', u'main-pane-height') {}
=== Setting Window Option ===
set_w_opt_dbg: "foo"
cmd_dbg1: set-window-option (u'-t$1:2', u'pane-border-format', u'foo') {}
cmd_dbg2: set-window-option (u'-t$1:2', u'pane-border-format', u'foo') {}
=== Showing Window Option ===
show_w_opt_dbg: here1
show_w_opt_dbg: here2
show_w_opt_dbg: here4 pane-border-format False
cmd_dbg1: show-window-options (u'pane-border-format',) {}
cmd_dbg2: show-window-options (u'-t', u'@2', u'pane-border-format') {}
=== Setting Window Option ===
set_w_opt_dbg: "#P"
cmd_dbg1: set-window-option (u'-t$1:2', u'pane-border-format', u'#P') {}
cmd_dbg2: set-window-option (u'-t$1:2', u'pane-border-format', u'#P') {}
=== Showing Window Option ===
show_w_opt_dbg: here1
show_w_opt_dbg: here2
show_w_opt_dbg: here4 pane-border-format False
cmd_dbg1: show-window-options (u'pane-border-format',) {}
cmd_dbg2: show-window-options (u'-t', u'@2', u'pane-border-format') {}
=== Setting Window Option ===
set_w_opt_dbg: " #P "
cmd_dbg1: set-window-option (u'-t$1:2', u'pane-border-format', u' #P ') {}
cmd_dbg2: set-window-option (u'-t$1:2', u'pane-border-format', u' #P ') {}
=== Showing Window Option ===
show_w_opt_dbg: here1
show_w_opt_dbg: here2
show_w_opt_dbg: here4 pane-border-format False
cmd_dbg1: show-window-options (u'pane-border-format',) {}
cmd_dbg2: show-window-options (u'-t', u'@2', u'pane-border-format') {}
======================================================== 1 failed in 0.47 seconds =========================================================
```
---

Also it probably needs to be

```py
    assert window.show_window_options('pane-border-format') == '" #P "'
```

instead of 

```py
    assert window.show_window_options('pane-border-format') == [' #P ']
```
?

**Comment by kaushalmodi** [comment]
I am at a deadend now.. In the failing case (`' #P '`), something is calling `Session._info`, but I cannot trace back any further.

https://github.com/kaushalmodi/libtmux/commit/1e10452e9d65780fcada05a1216c8c7242cd55b1

**Comment by tony** [comment]
I think it's understable that "format" type options such as:

- automatic-rename-format
- pane-border-format
- window-status-current-format
- window-status-format
- window-status-seperator

Are wrapped in quotes.

I think that these options should be unquoted on show, and re-quoted on setting.

**Comment by kaushalmodi** [comment]
> I think that these options should be unquoted on show, and re-quoted on setting.

That is not needed after part #2 of the fix; now comparison is with just `' #P '`.

And all the tests are now passing too!

---

.. though I still get this one failure. I don't get why Travis passes but it fails when I run `make test`.

```
_____________________________________________________________ test_set_height _____________________________________________________________

session = Session($1 libtmux_dBF9RL)

    def test_set_height(session):
        window = session.new_window(window_name='test_set_height')
        window.split_window()
        pane1 = window.attached_pane
        pane1_height = pane1['pane_height']

        pane1.set_height(4)
        assert pane1['pane_height'] != pane1_height
>       assert int(pane1['pane_height']) == 4
E       AssertionError: assert 3 == 4
E        +  where 3 = int('3')

tests/test_pane.py:36: AssertionError
================================================== 1 failed, 77 passed in 14.97 seconds ===================================================
```

**Comment by tony** [comment]
>\>       assert int(pane1['pane_height']) == 4
>E       AssertionError: assert 3 == 4
>E        +  where 3 = int('3')

@kaushalmodi Can I see your .tmux.conf?  How big is your terminal window? Have you tried making the font smaller/terminal window bigger so there are more cells?

File: tests/test_window.py [hunk 1]
```diff
@@ -174,13 +174,14 @@ def test_set_show_window_options(session):
...
     assert window.show_window_options('main-pane-height') == 20
 
-    window.set_window_option('pane-border-format', ' #P ')
-    assert window.show_window_options('pane-border-format') == [' #P ']
-
     window.set_window_option('main-pane-height', 40)
     assert window.show_window_options('main-pane-height') == 40
     assert window.show_window_options()['main-pane-height'] == 40
 
+    if has_gte_version('2.3'):
```
Comment (tony) [inline tests/test_window.py]: That was just added yesterday, looks like that function is already useful. :smile:

**Comment by kaushalmodi** [comment]
> Can I see your .tmux.conf? 

[here](https://github.com/kaushalmodi/dotfiles/blob/master/dot_tmux.conf)

I didn't realize that `make test` loaded the config too.

>How big is your terminal window? 

I think it is big enough
image

>Have you tried making the font smaller/terminal window bigger so there are more cells?

No

I thought got that error to go away if I moved my ~.tmux.conf. I can incrementally comment my config to see what's causing that. Thanks.

**Comment by kaushalmodi** [comment]
It's caused when the pane border is enabled!

> setw -g pane-border-status "bottom"

**Comment by tony** [comment]
Could you move your `.tmux.conf` then rerun the tests? No need to kill your main tmux server, since tests run on their own server.

Just trying to isolate if its a config issue or something else.

**Comment by kaushalmodi** [comment]
> Could you move your .tmux.conf then rerun the tests? 

That's what I did (see above).

> No need to kill your main tmux server, since tests run on their own server.

Yup.

**Comment by tony** [comment]
> I thought got that error to go away if I moved my ~.tmux.conf. I can incrementally comment my config to see what's causing that. Thanks.

Just read that, sorry. So did moving the tmux config make it test error go away?

**Comment by kaushalmodi** [comment]
> So did moving the tmux config make it test error go away?

Yes. That fixed the error, and also changing the `pane-border-status` to `"top"` instead of `"bottom"`.

**Comment by tony** [comment]
Okay that makes sense.  So yes I think we can say it's a config issue, for now, and that this issue is good to merge.

LGTM :shipit:

**Comment by kaushalmodi** [comment]
> So yes I think we can say it's a config issue

Seems like a tmux issue.

If I am in a tmux window with panes split into top/bottom, `display "#{pane_height}"` **increases** by 1 when I switch `pane-border-status` from `"top"` to `"bottom"`. 

Thanks for merging this PR.

**Comment by tony** [comment]
> If I am in a tmux window with panes split into top/bottom, display "#{pane_height}" increases by 1 when I switch pane-border-status from "top" to "bottom".

Understood.

> Thanks for merging this PR.

Thanks for your time and your contribution. It is now on pypi as libtmux 0.7.1. You are credited in the CHANGES.

I thought there as an issue in tmuxp you had, but I can't find it. I still need to add a test on the tmuxp side.

**Comment by kaushalmodi** [comment]
> I thought there as an issue in tmuxp you had, but I can't find it.

It is this same issue.. `tmuxp freeze` failed because I had set the pane status to `" #P "` (https://github.com/tony/tmuxp/issues/239).

**Pull request merged**